### ISSUE TITLE
Cairo WriteError while writing to output stream

### ISSUE THREAD

**Issue by JohnLudlow** [andymeneely/squib#236]
Hi,

I updated some things recently and now I get this error (this is a project I dip into occasionally):

```
Loading SVG(s) <===========================================> 100% Time: 00:00:00
Loading SVG(s) <===========================================> 100% Time: 00:00:00
Loading SVG(s) <===========================================> 100% Time: 00:00:00
rake aborted!
Cairo::WriteError: error while writing to output stream
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_pdf.rb:60:in `initialize'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_pdf.rb:60:in `new'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_pdf.rb:60:in `init_cc'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_pdf.rb:14:in `render_pdf'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/api/save.rb:22:in `save_pdf'
C:/Users/John/Documents/valar/rb/kingdom.rb:39:in `block in make_kingdom_deck'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/deck.rb:74:in `instance_eval'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/deck.rb:74:in `initialize'
C:/Users/John/Documents/valar/rb/kingdom.rb:4:in `new'
C:/Users/John/Documents/valar/rb/kingdom.rb:4:in `make_kingdom_deck'
C:/Users/John/Documents/valar/Rakefile:10:in `block in <top (required)>'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:251:in `block in execute'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:251:in `each'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:251:in `execute'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:195:in `block in invoke_with_call_chain'
C:/tools/ruby24/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:217:in `block in invoke_prerequisites'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:215:in `each'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:215:in `invoke_prerequisites'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:194:in `block in invoke_with_call_chain'
C:/tools/ruby24/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/task.rb:181:in `invoke'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:153:in `invoke_task'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:109:in `block (2 levels) in top_level'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:109:in `each'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:109:in `block in top_level'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:118:in `run_with_threads'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:103:in `top_level'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:81:in `block in run'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:179:in `standard_exception_handling'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/lib/rake/application.rb:78:in `run'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/rake-12.1.0/exe/rake:27:in `<top (required)>'
C:/tools/ruby24/bin/rake:23:in `load'
C:/tools/ruby24/bin/rake:23:in `<main>'
Tasks: TOP => default => spells
```

This is how I am compiling:
```
def make_kingdom_deck (prefix, deck, layout)
  Squib::Deck.new(cards: deck['Title'].size, layout: layout, width: 816, height: 1110) do
    background color: 'white'

    png file: deck['Art'], layout: :Art

    png file: 'wip/panel.png'
    png file: 'wip/bg-textile.png', blend: :soft_light

    text str: deck['Title'], layout: :Title
    text str: deck['Keywords'], layout: :Keywords


    text str: deck['Effect 1 title'], layout: :Effect1Title
    text str: deck['Effect 1 description'], layout: :Effect1Desc do |ctx|  
      ctx.svg key: ':gold:', file: "images/icons/coin.svg", layout: :CoinIcon
      ctx.png key: ':mana:', file: "images/icons/mana-mandatory.png", layout: :CoinIcon
    end

    text str: deck['Effect 2 title'], layout: :Effect2Title
    text str: deck['Effect 2 description'], layout: :Effect2Desc do |ctx|  
      ctx.svg key: ':gold:', file: "images/icons/coin.svg", layout: :CoinIcon
      ctx.png key: ':mana:', file: "images/icons/mana-mandatory.png", layout: :CoinIcon
    end

    svg file: 'icons/attack.svg', layout: :AttackIcon
    text str: deck['Attack'], layout: :AttackText

    svg file: 'icons/defend.svg', layout: :DefendIcon
    text str: deck['Defend'], layout: :DefendText

    svg file: 'icons/coin.svg', layout: :CoinIcon
    text str: deck['Gold Cost'], layout: :CoinText
    

    save_png prefix: prefix
    save_pdf trim: 36, file: prefix + ".pdf" #, gap: 10
  end
end

```
I commented out the save_png call to see if the save_pdf call worked, and it also failed in the same way.

This is called like this:
```
file :spells => ["Rakefile", "layouts/spells.yml", "rb/kingdom.rb", "decks/kingdom/spells.csv"] do
  spells = Squib.csv file: 'decks/kingdom/spells.csv'
  make_kingdom_deck 'spell_', spells, "layouts/spells.yml"
end
```

Versions:

Ruby version: `ruby 2.4.2p198 (2017-09-14 revision 59899) [x64-mingw32]`
Gem list:
```

*** LOCAL GEMS ***

bigdecimal (1.3.2, default: 1.3.0)
cairo (1.15.9 x64-mingw32)
cairo-gobject (3.1.8 x64-mingw32)
did_you_mean (1.1.2, 1.1.0)
gdk_pixbuf2 (3.1.8 x64-mingw32)
gio2 (3.1.8 x64-mingw32)
glib2 (3.1.8 x64-mingw32)
gobject-introspection (3.1.8 x64-mingw32)
io-console (default: 0.4.6)
json (2.1.0, default: 2.0.4)
mercenary (0.3.6)
mini_portile2 (2.2.0)
minitest (5.10.3, 5.10.1)
native-package-installer (1.0.4)
net-telnet (0.1.1)
nokogiri (1.8.0 x64-mingw32)
openssl (2.0.6, default: 2.0.5)
pango (3.1.8 x64-mingw32)
pkg-config (1.2.7)
power_assert (1.1.0, 0.4.1)
psych (2.2.4, default: 2.2.2)
rake (12.1.0, 12.0.0)
rdoc (5.1.0, default: 5.0.0)
roo (2.7.1)
rsvg2 (3.1.8 x64-mingw32)
ruby-progressbar (1.8.1)
rubyzip (1.2.1)
squib (0.13.4)
test-unit (3.2.6, 3.2.3)
xmlrpc (0.3.0, 0.2.1)
```

gemfile:

```
source 'https://rubygems.org'

gem 'squib'

gem 'guard'      # guard is what watches
gem 'guard-rake' # this hooks Guard into Rake
gem 'wdm', '>= 0.1.0' if Gem.win_platform? # optional, for watching directories
gem 'cairo'
```

The _output directory exists and can be written to. I am on Windows 10 x64 (I know, boo hiss and all that). I typically use PowerShell or VS Code's tasks, but this also happens from a cmd prompt as well.

If this is a case of something like cairo being updated in a breaking way, can you let me know what versions I need to specify in my gemfile?

Many thanks

**Comment by andymeneely** [comment]
Hi John - thanks for the bug report.

The only times I've gotten this error is when something else is trying to write to it at the same time. So I have a few questions on that:

  * Does this happen every time?
  * Does this still happen when you're not running under Guard?
  * What do you use to view the PDF?
  * Does this still happen when you comment out `save_pdf` and just do `save_png`?
  * I actually just started trying out VSCode (literally today...), and I noticed there were some weird Git issues on my setup that had to do with tracking the file changes and triggering those events. Perhaps try closing VSCode and any other Git clients? I have had rogue Git processes be spawned without my knowledge that are trying to index a file while being written. So may be try closing out as many of those things as you can. Or, if you're feeling up to it, you can search the file handle for the PDF on the Windows Resource Monitor

(btw I'm on Windows 10 x64 too... no complaints from me on that...)

Meanwhile, I'm only on Ruby 2.4.1p111, so I'll upgrade and see if I run into that issue.

Also, just for fun, try upgrading to the beta version that's out right now. Do `gem install squib --pre` and you'll get v0.14.beta1. Perhaps that changes things?

And finally, I don't see any issues with the code you wrote.

**Comment by JohnLudlow** [comment]
Thanks for your quick response. I'm at work now so will respond later with some more details.

> Does this happen every time?
Yes

> Does this still happen when you're not running under Guard?
I'll try removing this and let you know when I'm home

> What do you use to view the PDF?
Normally Chrome, though it was not open at the time - in fact the folder was empty.

> Does this still happen when you comment out save_pdf and just do save_png?
I'll try and let you know.

> Perhaps try closing VSCode and any other Git clients? 
I'll try and let you know.

I'll try running handle and let you know if there are any results.

**Comment by JohnLudlow** [comment]
Hi,

Here are the results of a couple of experiments

> Does this still happen when you're not running under Guard?
I wasn't really using guard because I could never quite get it working, so while it's referenced in the gemfile, it's not running at any point.

> Does this still happen when you comment out save_pdf and just do save_png?

Yes. 

I think the sample I gave and the output doesn't quite match, so apologies for that. The output seems to be from when I had commented out save_png.

Whichever of the save_xxx function calls I have first (and not commented out) is reported in the output with `--trace`.
```

Loading SVG(s) <===========================================> 100% Time: 00:00:00
Loading SVG(s) <===========================================> 100% Time: 00:00:00
Loading SVG(s) <===========================================> 100% Time: 00:00:00
rake aborted!o _output/spell__* <                            > 0% Time: 00:00:00
Cairo::WriteError: error while writing to output stream
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_images.rb:48:in `write_to_png'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_images.rb:48:in `write_png'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/graphics/save_images.rb:13:in `save_png'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/api/save.rb:31:in `block (2 levels) in save_png'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/args/card_range.rb:13:in `block in each'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/args/card_range.rb:13:in `each'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/args/card_range.rb:13:in `each'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/api/save.rb:30:in `block in save_png'
C:/tools/ruby24/lib/ruby/gems/2.4.0/gems/squib-0.13.4/lib/squib/progress.rb:29:in `start'
```

> Perhaps try closing VSCode and any other Git clients? 

Closed everything and same results

I tried using sysinternal's handle tool:

```
PS C:\Users\John\Documents\Valar> handle _output

Nthandle v4.1 - Handle viewer
Copyright (C) 1997-2016 Mark Russinovich
Sysinternals - www.sysinternals.com

No matching handles found.
PS C:\Users\John\Documents\Valar>
```

According to this nothing should be active in that folder. I did also try copying my code to another folder and running rake again, same results.

**Comment by andymeneely** [comment]
Do you call `make_kingdom_deck` multiple times? Maybe Squib is the one holding onto the handle.

**Comment by JohnLudlow** [comment]
Yes I do call it a few times, with differing parameters. I'll try tonight removing all but one call to `make_kingdom_deck`, and see what handles are in use before and after compilation.

This was working fine a few months ago (last time I dipped into this project seriously).

Also, if squib was holding onto the handle after writing a file, I would think that would imply some item would be in the folder. As in, the first item was written, but squib was still holding a handle to the folder so the second failed. In this case though, no output items are written at all.

Does squib use a temp folder to arrange things before adding them to the output folder?

I'll also take another look at procmon and see if anything jumps out. I did have a quick look with procmon before, but nothing jumped out. Perhaps a second look will help.

Thanks

**Comment by andymeneely** [comment]
> Does squib use a temp folder to arrange things before adding them to the output folder?

It does not. Everything is in memory until the very end. One exception to this is when `backend: svg` is set, but that wouldn't fail here.

> I decided to download and try the samples again, but I'm facing errors there too:

Getting the samples to work shouldn't require downloading the whole repo. If you do download the repo and want to work on Squib, then you'll need to do `bundle install` to install the development dependencies. The error you got about `bundler/gem_tasks` is from not having all the development deps installed I think.

> Also, if squib was holding onto the handle after writing a file, I would think that would imply some item would be in the folder. As in, the first item was written, but squib was still holding a handle to the folder so the second failed. In this case though, no output items are written at all.

Yeah so this is why I ended up asking the Cairo people what all can cause this error. Here's [the response](https://lists.cairographics.org/archives/cairo/2017-October/028428.html):

> - Trying to write a surface with width or height 0
> - When opening the output file fails (fopen())
> - When fwrite() to the output file fails, i.e. ferror() returns non-zero
> - When closing the output file fails (fclose())

One thing you can do, and this might get a tad involved but I would really appreciate it, is to 

1. Download the repo and install the latest version by just running `rake`. 
2. Make sure the problem is still happening.
3. In `lib/squib/graphics/save_pdf.rb`, put in `require 'irb'; binding.irb` right before that save call. 
4. Run your example again and you should get an IRB console just before the error. From there try a bunch of things. Try to open up a file, check the size of the Cairo surface, etc. Look for anything out of the ordinary.

I'd do this myself, but I can't seem to reproduce this on my own.

**Comment by JohnLudlow** [comment]
OK, I think I have solved it.

Bundle (a new one on me but I'm behind the times in Ruby terms) kept failing, complaining that it couldn't write to Gemfile.lock and this was probably a permissions issue. Now, I did reset my permissions on that folder (thinking it might have been a permissions issue) once before, to no effect.

I moved the folder somewhere else and reset the permissions again, and it now works. It's a little weird that I was able to create files in that folder from the powershell session I was using, but rake was failing to write stuff under the same session.

It may have been a weird profile management issue (Windows has opinions about permissions for user profiles). I generally don't keep source code under user profiles most of the time, but figured for this it would be OK.

Thanks again for your help.
